### PR TITLE
Support for Additional Profiles and Clusters

### PR DESCRIPTION
**Pull request by kylacochrane**
**This pull request is a work in progress and is not yet complete.**

This update aims to enhance the pipeline by integrating additional reference profiles and clusters from user-provided database parameters:

- `--db_profiles` will be incorporated through the `APPEND_PROFILES` process (which follows `LOCIDEX_MERGE_REF`).
- `--db_clusters` will be integrated via the `APPEND_CLUSTERS` process (which follows `CLUSTER_FILE`).

Both parameters are required for their respective processes, and users must provide both; it is not possible to supply only one.

Please note that tests are currently failing, and further tests will need to be implemented. 
No documentation has been updated yet. 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/phac-nml/gasnomenclature/tree/main/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -22,7 +22,15 @@ if (params.validate_params) {
     validateParameters()
 }
 
+/*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    Ensure both --db_profiles and --db_clusters are provided together
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
 
+if ((params.db_profiles && !params.db_clusters) || (!params.db_profiles && params.db_clusters)) {
```
Comment (emarinier) [inline main.nf:31]: You don't have to change this, because it's probably easier to understand the way you have it, but you could do this with an XOR: true only if one of the two is true.

I don't think Python has an actual XOR operator, so it might reduce to something like:

`if bool(params.db_profiles) != bool(params.db_clusters):`

but again, just a comment, not necessary to change.

**Review by emarinier — APPROVED** [review]
Looks good to me!

**Comment by sgsutcliffe** [comment]
It looks good nothing to change. Always have suggestions though! 

It could be made more clear in the README.md (or with additional error reporting but more work).  Basically, it should be emphasized that any address levels in the additional databases that are not in the samplesheet address will be dropped. The error could be triggered if max address size in the samplesheet is smaller than number of columns/levels in the database. As those will be will be dropped (I believe based on how csvtk concat works). For the README.md you could emphasize **headers must match** and do something like this for the example reference database. 
address 1.1. ... . n
| sample_id | l1  | l2  | ...| ln  |
| --------- | --- | --- | --- | --|
| sampleA   | 1   | 1   | ...   | 1
| sampleB   | 1   | 1   | ...   | 2
| sampleC   | 2   | 1   | ...   | 1

**Review by sgsutcliffe — APPROVED** [review]

File: nextflow_schema.json [hunk 2]
```diff
@@ -89,6 +89,27 @@
...
+        "databases": {
+            "title": "Databases",
+            "type": "object",
+            "description": "The optional databases of profiles and cluster addresses that can be used.",
+            "fa_icon": "fas fa-terminal",
+            "default": "",
+            "properties": {
+                "db_profiles": {
+                    "type": "string",
+                    "pattern": "^\\S+\\.tsv$",
```
Comment (apetkau) [inline nextflow_schema.json:101]: I am wondering if you could change this to allow both uncompressed `.tsv` and compressed `.tsv.gz` files as input? Since I would like to store the database files as gzipped tsv.

I know this will require a few additional changes outside of this file so that concatenation works properly (does csvtk work with gzipped files as well?)

File: tests/pipelines/main_append_databases.nf.test [hunk 3]
```diff
@@ -0,0 +1,91 @@
+nextflow_pipeline {
```
Comment (apetkau) [inline tests/pipelines/main_append_databases.nf.test:1]: Thanks so much for adding these test cases 😄

File: modules/local/append_clusters/main.nf [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    input:
+    path(initial_clusters)
+    path(additional_clusters)
+
+    output:
+    path("reference_clusters.tsv")
+
+    script:
+    """
+    csvtk concat -t ${initial_clusters} ${additional_clusters} > reference_clusters.tsv
```
Comment (apetkau) [inline modules/local/append_clusters/main.nf:18]: Do you think it would be possible to do some additional checking here to make sure all the clusters have addresses with the same number of levels? I don't know if there is some other csvtk command that could be used to check for this?

If it would be difficult, we could leave this out right now, but I would ideally like additional error checking done to make sure files are all consistent.

File: modules/local/append_profiles/main.nf [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    input:
+    path(reference_profiles)
+    path(additional_profiles)
+
+    output:
+    path("*.tsv")
+
+    script:
+    """
+    csvtk concat -t ${reference_profiles} ${additional_profiles} > profiles_ref.tsv
```
Comment (apetkau) [inline modules/local/append_profiles/main.nf:18]: Does `csvtk concat` make sure the columns all match (I forgot if it does)? What would happen if there are different loci in the `reference_profiles` and `additional_profiles`? I wonder if there's a way for us to check that?

Also, is there some way for us to check for e.g., an overlap of sample identifer in the `reference_profiles` and `additional_profiles`? Doesn't have to be in this process, but just something else I thought of for validation of the data.

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks so much for making these changes Kyla. Great work 😄 

A few in-line comments below

File: main.nf (hunk 1 above)
Comment (kylacochrane) [reply main.nf]: Great suggestion Eric - played around a bit and got it working here: [7d54942](https://github.com/phac-nml/gasnomenclature/pull/29/commits/7d549422c74403bf9ba8b08fcc6923d2396a175f)

File: nextflow_schema.json (hunk 2 above)
Comment (kylacochrane) [reply nextflow_schema.json]: Sure can! Luckily `csvtk` works with gzipped files: [28c23ae](https://github.com/phac-nml/gasnomenclature/pull/29/commits/28c23ae2845999803617ae1b92cd7ee2914c2fe1)
The corresponding test data files were added: [bb91d0d](https://github.com/phac-nml/gasnomenclature/pull/29/commits/bb91d0dd393384273666a826e60549152d61baeb)
I did have to add some syntax to deal with gzipped files when doing a header check with the `append` modules, which can be found here: [3002b00](https://github.com/phac-nml/gasnomenclature/pull/29/commits/3002b00b21dac984ebc1c71d2a886f4cd1974935)

File: modules/local/append_clusters/main.nf (hunk 4 above)
Comment (kylacochrane) [reply modules/local/append_clusters/main.nf]: I think I figured out a way to do this, but I am going to save it for the next PR as I will be making changes to the formatting of the `--db_clusters` and `cluster_file` modules to align with the changes made to `GAS CALL`.

File: modules/local/append_profiles/main.nf (hunk 5 above)
Comment (kylacochrane) [reply modules/local/append_profiles/main.nf]: No, `csvtk` does not have a way to ensure all columns match.
Instead I added a function and a check to the script in both `append_clusters` and `append_profiles` here: [8ae527f](https://github.com/phac-nml/gasnomenclature/pull/29/commits/8ae527f185c3e89de9e5f9b2d7e0b8ca88ea6356)
I also used `csvtk uniq` to ensure there are no overlapping `sample_id's` : [8ae527f](https://github.com/phac-nml/gasnomenclature/pull/29/commits/8ae527f185c3e89de9e5f9b2d7e0b8ca88ea6356)

**Comment by kylacochrane** [comment]
Thank you everyone for your review and suggestions! 
I have updated the `README` but additional changes will need to be made once the new version of `GAS CALL` is used, as there will be simplified formatting introduced.

File: modules/local/append_clusters/main.nf (hunk 4 above)
Comment (apetkau) [reply modules/local/append_clusters/main.nf]: Thanks so much for this Kyla 😄

File: modules/local/append_profiles/main.nf [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    add_headers=\$(get_header "${additional_profiles}")
+
+    if [ "\$ref_headers" != "\$add_headers" ]; then
+        echo "Error: Column headers do not match between reference_profiles and --db_profiles."
+        exit 1
+    fi
+
+    # Merge profiles ensuring only unique samples are added
+    csvtk concat -t ${reference_profiles} ${additional_profiles} > profiles.tsv
+    csvtk uniq -t -f sample_id profiles.tsv > profiles_ref.tsv
```
Comment (apetkau) [inline modules/local/append_profiles/main.nf:38]: Thanks so much for adding this in to check for and remove duplicate samples.

One question I have, the docs for csvtk (https://bioinf.shenwei.me/csvtk/usage/#uniq) say that it applies unique without sorting. Does this mean that it will only remove duplicate samples if they happen right next to each other? Could you check?

File: modules/local/append_profiles/main.nf (hunk 5 above)
Comment (apetkau) [reply modules/local/append_profiles/main.nf]: Thanks so much Kyla. This is amazing work 😄

**Review by apetkau — COMMENTED** [review]
Thanks so much Kyla for addressing all my comments. And for adding all those tests 😄. Amazing work.

I just have one more question given in-line.

File: modules/local/append_profiles/main.nf (hunk 6 above)
Comment (kylacochrane) [reply modules/local/append_profiles/main.nf]: Great catch, Aaron! You're absolutely right. I've added a `csvtk sort` to ensure duplicate sample_ids are grouped correctly for proper removal by `csvtk uniq`. I also updated the tests to account for the sorting and adjusted the order of the duplicate sample_ids so they are no longer adjacent.
Changes can be found here: [9780166](https://github.com/phac-nml/gasnomenclature/pull/29/commits/9780166b3250c45773f2bde5f720fe9f391ea488)
And test updates here: [7d2f30a](https://github.com/phac-nml/gasnomenclature/pull/29/commits/7d2f30ab31dafefcbe643ace3c136c73813cb3c1)

File: modules/local/append_clusters/main.nf [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    csvtk mutate2 -t -n source -e " 'ref' " ${initial_clusters} > reference_clusters_source.tsv
+    csvtk mutate2 -t -n source -e " 'db' " ${additional_clusters} > additional_clusters_source.tsv
+
+    # Combine profiles from both the reference and database into a single file
+    csvtk concat -t reference_clusters_source.tsv additional_clusters_source.tsv | csvtk sort -t -k id > combined_profiles.tsv
+
+    # Calculate the frequency of each sample_id across both sources
+    csvtk freq -t -f id combined_profiles.tsv > sample_counts.tsv
+
+    # For any sample_id that appears in both the reference and database, add a 'db_' prefix to the sample_id from the database
```
Comment (apetkau) [inline modules/local/append_clusters/main.nf:46]: That's really cool. This would solve the issue with duplicates in all situations then 😄 . Thanks so much.

File: tests/pipelines/main_append_databases.nf.test [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+            assert actual_cluster.text == expected_cluster.text
+
+            // Check called clusters
+            def actual_calls = path("$launchDir/results/call/Called/results.text")
+            def expected_calls = path("$baseDir/tests/data/called/expected_results_append.txt")
+            assert actual_calls.text == expected_calls.text
+        }
+    }
+
+    test("Test pipeline when appended profiles or clusters have sample_id overlap") {
```
Comment (apetkau) [inline tests/pipelines/main_append_databases.nf.test:125]: Thank you for adding this test 😄

**Review by apetkau — APPROVED** [review]
Thanks so much for all the great work you've done with this Kyla. It looks amazing and handles so many more situations with sample names. I really appreciate it 😄

File: modules/local/append_profiles/main.nf (hunk 6 above)
Comment (apetkau) [reply modules/local/append_profiles/main.nf]: Thank you for all the great work you've put into this :)
